### PR TITLE
Add script for OptimPack

### PR DESCRIPTION
**Pull request by emmt**
This is a new script to build [OptimPack](https://github.com/emmt/OptimPack) C libraries to be used in Julia via the [OptimPack.jl](https://github.com/emmt/OptimPack.jl) package.

### CODE REVIEW

**Comment by giordano** [comment]
Thanks for the contribution!  Out of curiosity, were you able to build for all platforms? If so, which version of BinaryBuilder did you use?

File: O/OptimPack/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/optimpack-*
+./configure --prefix=$prefix --host=$target --enable-shared=yes --enable-static=no
+FLAGS=()
+if [[ "${target}" == *-mingw* ]]; then
+    # This is needed in order to build the shared library on Windows
+    FLAGS+=(LDFLAGS="-no-undefined")
```
Comment (giordano) [inline O/OptimPack/build_tarballs.jl]: I updated the script to make it compatible with `master` version of BinaryBuilder.

When building for Windows, I was getting this error:
```
libtool:   error: can't build i686-w64-mingw32 shared library unless -no-undefined is specified
```
I simply followed the error message and added the `-no-undefined` flag for the linker, but I'm not convinced this is the right solution here :thinking:  At least it makes it possible to build the project

Comment (emmt) [reply O/OptimPack/build_tarballs.jl]: Thanks for the fix. I do not understand myself the meaning of this flag. Let us give it a try.

**Comment by emmt** [comment]
> Thanks for the contribution! Out of curiosity, were you able to build for all platforms? If so, which version of BinaryBuilder did you use?

I used BinaryBuilder version v0.1.4 from Julia 1.2. I have seen that BonaryBuilder 0.1.4 is pretty old compared to master (last modifcation in the src dir. was 10 months ago). Shall I switch to the master version of BinaryBuilder and write scripts compatible for this version to contribute to Yggdrasil?

**Comment by giordano** [comment]
> I used BinaryBuilder version v0.1.4 from Julia 1.2.

So with this version you were able to build for all platforms?

> Shall I switch to the master version of BinaryBuilder and write scripts compatible for this version to contribute to Yggdrasil?

Master version of BinaryBuilder requires Julia 1.3rc4, but it's not a problem if you cannot use it, the changes necessary to adapt to this version are very simple.

**Comment by emmt** [comment]
> > I used BinaryBuilder version v0.1.4 from Julia 1.2.
> 
> So with this version you were able to build for all platforms?

No, in fact I just followed the instructions in [Yggdrasil](https://github.com/JuliaPackaging/Yggdrasil) and built for the 2 Linux platforms (64 and 32 bits). My understanding was that I did not have to install all the docker files needed by BinaryBuilder, just to check that the script worked so I tested that with:

```sh
julia --color=yes build_tarballs.jl x86_64-linux-gnu i686-linux-gnu --verbose --debug
```

where `x86_64-linux-gnu` `i686-linux-gnu` were the 2 compilers available on my machine (LinuxMint 19.2 Tina). For such a simple project as OptimPack (no other dependencies than the sin/cos/sqrt functions in the math library and the calloc/malloc/free/printf in the standard C library), I did not anticipate that it would be necessary to check building for all other platforms (Windows, MacOS, etc.)

> > Shall I switch to the master version of BinaryBuilder and write scripts compatible for this version to contribute to Yggdrasil?
> 
> Master version of BinaryBuilder requires Julia 1.3rc4, but it's not a problem if you cannot use it, the changes necessary to adapt to this version are very simple.

OK but even if I do not have to install the very last master version of BinaryBuilder, to check building for all other platforms I have to install all the docker files needed by BinaryBuilder.

In that case, why not just upload myself the built libraries on the GitHub release page of the project (and this can be automated by TravisCI) or is there something I have missed about the purpose of Yggdrasil?

**Comment by giordano** [comment]
> the 2 compilers available on my machine (LinuxMint 19.2 Tina)

I'm confused by this comment.  BinaryBuilder provides you an environment where to cross-compile a library for 13 different platforms without you having to install anything on your machine.

> to check building for all other platforms I have to install all the docker files needed by BinaryBuilder.

You said that you're using GNU/Linux, then you don't need Docker.  The default option is to use UserNS.

> In that case, why not just upload myself the built libraries on the GitHub release page of the project (and this can be automated by TravisCI) or is there something I have missed about the purpose of Yggdrasil?

The goal of Yggdrasil is to collect all (or the vast majority of) builders for libraries compiled with BinaryBuilder.  The binaries are uploaded to relevant repository under https://github.com/JuliaBinaryWrappers.  These are JLL packages, the new way to provide binary libraries that will go live with Julia 1.3.  The binaries themselves can still be installed in Julia <=1.2 with BinaryProvider.

**Comment by emmt** [comment]
OK I've just tried to run the `build_tarballs.jl` script (modified a bit to work with BinaryBuilder v0.1.4) with no restricted list of targets and interactively to have BinaryBuilder download the Apple SDK. That's awesome: I now have the compiled libraries for 12 different architectures. I also have a gigantic (~13Gb) `~/.julia/packages/BinaryBuilder` directory but that's a minor drawback.  

Then I switched to the master version of BinaryBuilder:

```julia
(...) pkg> dev BinaryBuilder
```

and run the `build_tarballs.jl` script as modified by you interactively from Julia 1.3rc4 (as you told me the latest version of Julia is needed), I just had to fix a bug in `macos_sdk_already_installed()` (see PR https://github.com/emmt/BinaryBuilder.jl/tree/patch-1) to have it produce the 12 versions of the libraries. The packages were bigger now because compiled with `-O2 -g` (no optimisation by default with BinaryBuilder v0.1.4) and it is the  `~/.julia/artifacts/` directory which weights 16Gb now. Again not a big deal.

Now I think I better understand all this. You have been very helpful, many thanks!

**Comment by staticfloat** [comment]
Great work!  Thanks!

**Pull request merged**